packages/desktop-client/src/components/accounts/TransactionsTable.js

```diff
@@ -1787,6 +1837,120 @@ export let TransactionTable = forwardRef((props, ref) => {
+  let [ascDesc, setAscDesc] = useState('desc');
+
+  function onSortTable(e) {
+    let transform =
+      e.currentTarget.innerText.toLowerCase() === 'payment' ||
+      e.currentTarget.innerText.toLowerCase() === 'deposit'
+        ? 'amount'
+        : e.currentTarget.innerText.toLowerCase();
+    if (transform === field) {
+      ascDesc === 'desc' ? setAscDesc('asc') : setAscDesc('desc');
+    } else {
+      setField(transform);
+      transform === 'amount' || transform === 'date'
+        ? setAscDesc('asc')
+        : setAscDesc('desc');
+    }
+  }
+
+  let category = props.categoryGroups.flatMap(cat => {
+    return cat.categories;
+  });
+  let fields = [field] + 's';
+
+  function sortStrings() {
+    transactions =
+      ascDesc === 'asc'
+        ? transactions
+            .map(e => {
+              return {
+                ...e,
+                test: props[fields].find(f => {
+                  return f.id === e[field];
+                })
+                  ? props[fields].find(f => {
+                      return f.id === e[field];
+                    }).name
+                  : '',
+              };
+            })
+            .sort((a, b) => a.test < b.test)
+        : transactions
+            .map(e => {
+              return {
+                ...e,
+                test: props[fields].find(f => {
+                  return f.id === e[field];
+                })
+                  ? props[fields].find(f => {
+                      return f.id === e[field];
+                    }).name
+                  : '',
+              };
+            })
+            .sort((a, b) => a.test > b.test);
+  }
+
+  function sortCat() {
+    transactions =
+      ascDesc === 'asc'
+        ? transactions
+            .map(e => {
+              return {
+                ...e,
+                test: category.find(f => {
+                  return f.id === e[field];
+                })
+                  ? category.find(f => {
+                      return f.id === e[field];
+                    }).name
+                  : '',
+              };
+            })
+            .sort((a, b) => a.test < b.test)
+        : transactions
+            .map(e => {
+              return {
+                ...e,
+                test: category.find(f => {
+                  return f.id === e[field];
+                })
+                  ? category.find(f => {
+                      return f.id === e[field];
+                    }).name
+                  : '',
+              };
+            })
+            .sort((a, b) => a.test > b.test);
+  }
+
+  if (field) {
+    switch (field) {
+      case 'account':
+        sortStrings();
+        break;
+      case 'payee':
+        sortStrings();
+        break;
+      case 'category':
+        sortCat();
+        break;
+      case 'date':
+        ascDesc === 'asc'
+          ? transactions.sort((a, b) => a[field] < b[field])
+          : transactions.sort((a, b) => a[field] > b[field]);
```

Comment:
All of this is mutating the `transactions` array which is undesirable IMO — especially since it runs on every render, which could potentially be slow if looking at a large list of transactions. Could you move it into a `useMemo` that returns a new array so that it only needs to run if one of the relevant props change? That would also make it possible to factor it out into a custom `useSortedTransactions` hook that could be used in other transaction tables, potentially.

Reply:
Makes sense, I'll make these changes.